src/python/pants/notes/master.rst

```diff
@@ -4,6 +4,140 @@ Master Pre-Releases
+* Fix `validate` and `cloc2` breaking with targets with no sources (#9000)
+  `PR #9000 <https://github.com/pantsbuild/pants/pull/9000>`_
+
+* Fix rule type hint validation to work with `from __future__ import annotations` (#9004)
+  `PR #9004 <https://github.com/pantsbuild/pants/pull/9004>`_
+
+* correctly fingerprint scoverage options and copy output into dist/ (#8705)
+  `PR #8705 <https://github.com/pantsbuild/pants/pull/8705>`_
+
+* Restore support for `dependencies` outputting 3rdparty deps through new enum option (#8960)
+  `PR #8960 <https://github.com/pantsbuild/pants/pull/8960>`_
+
+* Set workunit parent IDs correctly (fixes #7969) (#8947)
+  `PR #8947 <https://github.com/pantsbuild/pants/pull/8947>`_
+
+Refactoring, Improvements, and Tooling
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+* Fix some clippy errors which will appear in 1.41.0 (#9027)
+  `PR #9027 <https://github.com/pantsbuild/pants/pull/9027>`_
+
+* Expand `config_test.py` to better define the Config API (#9021)
+  `PR #9021 <https://github.com/pantsbuild/pants/pull/9021>`_
+
+* Refactor `engine/legacy/structs.py` (#9023)
+  `PR #9023 <https://github.com/pantsbuild/pants/pull/9023>`_
+
+* Describe the origin of failure when globs do not match (#9013)
+  `PR #9013 <https://github.com/pantsbuild/pants/pull/9013>`_
+
+* Small improvements to release script (#8964)
+  `PR #8964 <https://github.com/pantsbuild/pants/pull/8964>`_
+
+* First usage of async-await (#9018)
+  `PR #9018 <https://github.com/pantsbuild/pants/pull/9018>`_
+
+* Refactor `config.py` to allow supporting possible future config file formats (#9014)
+  `PR #9014 <https://github.com/pantsbuild/pants/pull/9014>`_
+
+* Bump to Rust 1.40 (#9017)
+  `PR #9017 <https://github.com/pantsbuild/pants/pull/9017>`_
+
+* Improve unmatched globs error message (#9010)
+  `PR #9010 <https://github.com/pantsbuild/pants/pull/9010>`_
+
+* Improve error messages for rule graph issues (#8998)
+  `PR #8998 <https://github.com/pantsbuild/pants/pull/8998>`_
+
+* Add type hints to `options.py`, `options_bootstrapper.py`, and `config.py` (#8990)
+  `PR #8990 <https://github.com/pantsbuild/pants/pull/8990>`_
+
+* Add IPython subsystem and deprecate `--repl-py` options (#8996)
+  `PR #8996 <https://github.com/pantsbuild/pants/pull/8996>`_
+
+* Move intrinsics code into new module (#8999)
+  `PR #8999 <https://github.com/pantsbuild/pants/pull/8999>`_
+
+* Change distribution lookup for JvmTask subclasses to call into JvmTask (#8938)
+  `PR #8938 <https://github.com/pantsbuild/pants/pull/8938>`_
+
+* Refactor `--changed-include-dependees` to use an Enum option (#8995)
+  `PR #8995 <https://github.com/pantsbuild/pants/pull/8995>`_
+
+Testing
+~~~~~~~
+
+* Fix Zipp 2.0 release breaking Python 2 integration tests (#8994)
+  `PR #8994 <https://github.com/pantsbuild/pants/pull/8994>`_
+
+Documentation
+~~~~~~~~~~~~~
+* Register certain optional conditionally on v1/v2 setup. (#8997)
+  `PR #8997 <https://github.com/pantsbuild/pants/pull/8997>`_
```

Comment:
I think either an API change or refactoring, but not documentation. This impacts actual users.